### PR TITLE
8278518: String(byte[], int, int, Charset) constructor and String.translateEscapes() miss bounds check elimination

### PR DESCRIPTION
**Pull request by rwestrel**
In the micro benchmark for this bug, the bytecode has a single loop
head with 2 backedges. One branch is taken a lot more frequently than
the other (for instance, profiling reports 3820938 for the most
common, 134068 for the least common). When ciTypeFlow builds its loop
tree, the least common loop ends up as the inner loop (ciTypeFlow uses
the tail block number to order loops with a shared header). Then
ciTypeFlow clones the inner loop head.

The results I get then is:

StringBenchmark.safeDecoding    avgt    5  141.965 ± 53.511  ns/op
StringBenchmark.unsafeDecoding  avgt    5  123.051 ± 77.855  ns/op

The unsafe version uses unsafe instead of standard array accesses. The
reason for the unsafe version is to demonstrate that a missed range
check elimination in the common path hurts performance (the range
check is not performed on the loop iv but another phi so can't be
eliminated).

The patch I propose takes back branch count (from profile data) into
account when building the ciTypeFlow tree. As a consequence, the inner
loop (the one for which the loop head is cloned), is the most most
frequent one. The range check in the common path is also hoisted in
that case (it's now performed on the iv). Running the micro benchmark
again, I get:

StringBenchmark.safeDecoding    avgt    5  53.368 ± 2.232  ns/op
StringBenchmark.unsafeDecoding  avgt    5  55.565 ± 5.124  ns/op

Both are improved (2-3x) and the unsafe version doesn't perform better.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278518](https://bugs.openjdk.java.net/browse/JDK-8278518): String(byte[], int, int, Charset) constructor and String.translateEscapes() miss bounds check elimination

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [John R Rose](https://openjdk.java.net/census#jrose) (@rose00 - **Reviewer**) ⚠️ Review applies to 92b5d2d31589aca26718b57dbe58e24da2336b34

### Contributors
 * Sergey Tsypanov `<stsypanov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7034/head:pull/7034` \
`$ git checkout pull/7034`

Update a local copy of the PR: \
`$ git checkout pull/7034` \
`$ git pull https://git.openjdk.java.net/jdk pull/7034/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7034`

View PR using the GUI difftool: \
`$ git pr show -t 7034`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7034.diff">https://git.openjdk.java.net/jdk/pull/7034.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/ci/ciTypeFlow.cpp [hunk 1]
```diff
@@ -2459,20 +2526,26 @@ void ciTypeFlow::PreorderLoops::next() {
...
   Loop* current = leaf;
   while (lp != NULL) {
     int lp_pre_order = lp->head()->pre_order();
+    int lp_count = profiled_count(method, lp);
     // Find insertion point for "lp"
     while (current != NULL) {
       if (current == lp)
         return leaf; // Already in list
       if (current->head()->pre_order() < lp_pre_order)
         break;
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciTypeFlow.cpp:2541]: Can you add `{}` for these checks too.

**Review by vnkozlov — COMMENTED** [review]
Good. Can you add JMH benchmark for this case?

File: src/hotspot/share/ci/ciTypeFlow.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/ci/ciTypeFlow.cpp]: fixed in new commit

**Comment by rwestrel** [comment]
> Good. Can you add JMH benchmark for this case?

Thanks for the review.
I'm not the author of the benchmark. That actually came up in another PR where a library fix was discussed:
https://github.com/openjdk/jdk/pull/6812
Shouldn't the benchmark's author (that is @stsypanov or @amirhadadi) submit it (and sign the OCA if required)?

**Comment by stsypanov** [comment]
@rwestrel The benchmark is mine and OCA is signed. Here's the code
```java
@State(Scope.Thread)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class StringConstructorBenchmark {
  private byte[] array;
  private String str;

  @Setup
  public void setup() {
    str = "Quizdeltagerne spiste jordbær med fløde, mens cirkusklovnen. Я";//Latin1 ending with Russian
    array = str.getBytes(StandardCharsets.UTF_8);
  }

  @Benchmark
  public String newString()  {
      return new String(array, 0, array.length, StandardCharsets.UTF_8);
  }

  @Benchmark
  public String translateEscapes()  {
      return str.translateEscapes();
  }
}
```

**Comment by amirhadadi** [comment]
@stsypanov @rwestrel since the fix is general and does not target String constructor or translateEscapes directly, maybe the benchmark should explicitly test code that has the shape being optimized in this fix? i.e. a loop with 2 backedges, the second one being less frequent.

**Comment by vnkozlov** [comment]
Thank you, @stsypanov for clarifying status of your benchmark.
@amirhadadi I agree that to have simple benchmark is good to have. But I also want to add Sergey's benchmark because it shows the issue in this bug description.  I suggest to add both benchmarks.

**Comment by vnkozlov** [comment]
I run testing and in tier3 our `Kitchensink` closed testing failed:
```
#  Internal Error (/workspace/open/src/hotspot/share/ci/ciTypeFlow.cpp:2497), pid=15398, tid=15431
#  Error: ShouldNotReachHere()

Stack: [0x00007fd0fcf68000,0x00007fd0fd069000],  sp=0x00007fd0fd0657e0,  free space=1013k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x9cc2e7]  profiled_count(ciMethod*, ciTypeFlow::Loop*)+0x537
V  [libjvm.so+0x9cc4dd]  ciTypeFlow::Loop::sorted_merge(ciTypeFlow::Loop*, ciMethod*)+0x4d
V  [libjvm.so+0x9cff00]  ciTypeFlow::build_loop_tree(ciTypeFlow::Block*)+0x1c0
V  [libjvm.so+0x9d1b32]  ciTypeFlow::df_flow_types(ciTypeFlow::Block*, bool, ciTypeFlow::StateVector*, ciTypeFlow::JsrSet*)+0x632
V  [libjvm.so+0x9d208e]  ciTypeFlow::flow_types()+0x41e
V  [libjvm.so+0x9d2bb6]  ciTypeFlow::do_flow()+0x26
V  [libjvm.so+0x965b24]  ciMethod::get_flow_analysis()+0x64
V  [libjvm.so+0x7473ae]  InlineTree::check_can_parse(ciMethod*)+0xee
V  [libjvm.so+0x8b7579]  CallGenerator::for_inline(ciMethod*, float)+0x19
V  [libjvm.so+0xa999d4]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x12e4
```

**Comment by vnkozlov** [comment]
I added bytecode print and tested: `invalid bytecode: goto_w`

**Comment by vnkozlov** [comment]
I suggest to replace `ShouldNotReachHere()` with `fatal(" invalid bytecode: %s", Bytecodes::name(iter.cur_bc()));`

**Comment by rwestrel** [comment]
> I suggest to replace `ShouldNotReachHere()` with `fatal(" invalid bytecode: %s", Bytecodes::name(iter.cur_bc()));`

Thanks for running testing and the suggestion. Should be fixed now.

**Comment by rwestrel** [comment]
> @rwestrel The benchmark is mine and OCA is signed. Here's the code

I added the benchmark. What copyright do you want on it?

**Comment by stsypanov** [comment]
@rwestrel it could be anything

**Comment by rwestrel** [comment]
> @rwestrel it could be anything
> 
> P.S. as of `@contributor` command, try with my email: [sergei.tsypanov@yandex.ru](mailto:sergei.tsypanov@yandex.ru)

Ok. Thanks.

**Comment by rose00** [comment]
This is a good improvement.  I agree that the type-flow pass "plays dumb" about graph structure; it was originally designed even "dumber".  The idea was that following phases could arrange loop structure all by themselves. But this gave up some early optimizations in the parsing phase, which is too early for loop inference.  So type-flow has gotten "smarter" over the years, to give the parsing phase more foresight about loop structure.  This new improvement is appropriate to that design.

As a matter of style and readability I suggest threading the `ciMethod` pointer (which is the root for getting profile info out) through instance variables rather than random extra arguments.  The problem with random extra arguments is they could be any (random) `ciMethod` but we know it's only the statically relevant method to the whole `ciTypeFlow` pass.  You can't tell from reading the code (except as a whole) what that random argument is, but you *can* if the `ciMethod` in question is the one stored in `ciTypeFlow::_method`.

So, I'd use the inner/outer class design pattern, where `Loop` gets an `_outer` link that points to the containing `ciTypeFlow`, which in turn has the `_method` field you need.  Just add a constructor argument to Loop to set up the `_outer` link and you don't need the ad hoc extra argument to `sorted_merge`.

I notice that you cleaned up `clone_loop_heads` in the same way, bravo.

I don't fully follow the open-coded sorting logic in `sorted_merge`, and I wish it could be made more clear.  The expression `current_count == lp_count` is particularly puzzling to me; either it should be a `<=` or `>=` comparison, or it is some sort of tie-breaking logic.  The role of all of those comparisons would be easier to follow if there were (I don't know if it's possible?) a separately factored (partial) ordering function, as a static helper function.  The problem with open-coding the comparison logic in an insertion sort algorithm, as you have here, is that if you have multiple sorting keys, it's really hard to verify that they are correctly applied.

Also (minor nit) you use the pattern `if (pred1) { break; }; if (pred2) { break; }` and also the equivalent pattern `if (pred1) { break; } else if (pred2) { break; }`.  I prefer the first, and would find the code more readable if you just picked one or the other in any case.

**Comment by rwestrel** [comment]
Thanks for the review @rose00. What about the updated change?

**Review by rose00 — APPROVED** [review]
Thank you; that's much clearer.

Reviewed!

A couple of minor style issues I noticed which you might address (or not):

- The name insertion_point should be probably be named is_insertion_point or at_insertion_point  since it's a predicate
- In `assert((succs->at(ciTypeFlow::IF_NOT_TAKEN) == loop->head()) == (tail->limit() == loop->head()->start()), "bytecode and CFG not consistent")` the top-level `==` should be reversed so that the assert looks more like the previous assert.  It's easier to tell what they are doing if their structures line up with each other.

I wish we could handle profile counts on `switch` arms.  But that's a matter for another day.

**Comment by rwestrel** [comment]
> Thank you; that's much clearer.
> 
> Reviewed!

Thanks

> A couple of minor style issues I noticed which you might address (or not):
> 
>     * The name insertion_point should be probably be named is_insertion_point or at_insertion_point  since it's a predicate
> 
>     * In `assert((succs->at(ciTypeFlow::IF_NOT_TAKEN) == loop->head()) == (tail->limit() == loop->head()->start()), "bytecode and CFG not consistent")` the top-level `==` should be reversed so that the assert looks more like the previous assert.  It's easier to tell what they are doing if their structures line up with each other.

Done.

> I wish we could handle profile counts on `switch` arms. But that's a matter for another day.

The other blind spot is if the tail falls through to the head (and there's no if at the end of the tail to branch out of the loop).

**Comment by rwestrel** [comment]
@vnkozlov can you take another look at this now that John reviewed it?

**Review by vnkozlov — COMMENTED** [review]
Looks good. Let me test it before approval.

**Comment by vnkozlov** [comment]
I got an other failure in `Kitchecksink`:
```
#  Internal Error (/workspace/open/src/hotspot/share/interpreter/bytecode.cpp:84), pid=5553, tid=18400
#  assert(have_fmt == need_fmt) failed: assert_offset_size
#
# JRE version: Java(TM) SE Runtime Environment (19.0) (fastdebug build 19-internal+0-2022-01-25-1858238.vladimir.kozlov.jdkgit)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 19-internal+0-2022-01-25-1858238.vladimir.kozlov.jdkgit, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, serial gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x73de5e]  Bytecode::assert_offset_size(int, Bytecodes::Code, bool)+0xee
#

Current CompileTask:
C2:1220999 49982       4       die.verwandlung.xmlspec::applyTemplates (2549 bytes)

Stack: [0x00007feb068e9000,0x00007feb069ea000],  sp=0x00007feb069e6760,  free space=1013k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x73de5e]  Bytecode::assert_offset_size(int, Bytecodes::Code, bool)+0xee
V  [libjvm.so+0x79d05e]  ciBytecodeStream::get_dest() const+0x9e
V  [libjvm.so+0x9cc300]  ciTypeFlow::profiled_count(ciTypeFlow::Loop*)+0x3c0
V  [libjvm.so+0x9ccb50]  ciTypeFlow::Loop::sorted_merge(ciTypeFlow::Loop*)+0x250
V  [libjvm.so+0x9d036c]  ciTypeFlow::build_loop_tree(ciTypeFlow::Block*)+0x1bc
V  [libjvm.so+0x9d1f72]  ciTypeFlow::df_flow_types(ciTypeFlow::Block*, bool, ciTypeFlow::StateVector*, ciTypeFlow::JsrSet*)+0x632
V  [libjvm.so+0x9d24ce]  ciTypeFlow::flow_types()+0x41e
V  [libjvm.so+0x9d2ff6]  ciTypeFlow::do_flow()+0x26
V  [libjvm.so+0x965cd4]  ciMethod::get_flow_analysis()+0x64
V  [libjvm.so+0x7478fe]  InlineTree::check_can_parse(ciMethod*)+0xee
V  [libjvm.so+0x8b7549]  CallGenerator::for_inline(ciMethod*, float)+0x19
V  [libjvm.so+0xa99ca4]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x12e4
```

**Comment by rwestrel** [comment]
@vnkozlov thanks for the review and testing.
I pushed one more commit which I expect will fix that issue but I can't test it.

**Comment by vnkozlov** [comment]
I submitted new testing.

**Review by vnkozlov — APPROVED** [review]
Finally testing is clean.

**Comment by rwestrel** [comment]
> Finally testing is clean.

Thanks!

**Pull request closed without merging**